Kernel checker

bernardjkim opened the pull request:
### Description
Add a kernel checker that checks if the currently installed kernel version is supported by Gravity.

### Linked tickets and PRs
* Updates https://github.com/gravitational/gravity/issues/1818
* Requires https://github.com/gravitational/satellite/pull/248

bernardjkim committed:
Kernel checker

bernardjkim committed:
Include patch or build number

a-palchikov commented on monitoring/boot_config_linux.go:
```diff
@@ -110,31 +110,35 @@ func KernelVersionLessThan(version KernelVersion) KernelConstraintFunc {
 			(testVersion.Release == version.Release &&
```
```go
func KernelVersionLessThan(version KernelVersion) KernelConstraintFunc {
	return func(testVersion KernelVersion) bool {
		release := testVersion.Release <= version.Release
		major := testVersion.Major <= version.Major
		minor := testVersion.Minor <= version.Minor
		patch := testVersion.Patch < version.Patch
		return release && major && minor && patch
	}
}
```
otherwise `3.1.10-1125` < `3.0.10-1126`

bernardjkim replied:
Thanks for catching this. I've added a few extra test cases.

r0mant commented on monitoring/boot_config_linux.go:
```diff
@@ -110,27 +110,35 @@ func KernelVersionLessThan(version KernelVersion) KernelConstraintFunc {
+	Patch int
+}
+
+func (r *KernelVersion) String() string {
```
Nit: Godoc.

r0mant commented on monitoring/boot_config_linux.go:
```diff
@@ -230,9 +238,10 @@ type kernelVersionReader func() (version string, err error)
 type bootConfigReader func(release string) (io.ReadCloser, error)
 
 func parseKernelVersion(input string) (*KernelVersion, error) {
```
Could you add a comment on this function explaining what format of kernel version string it expects?

r0mant commented on monitoring/boot_config_linux.go:
```diff
@@ -253,10 +262,18 @@ func parseKernelVersion(input string) (*KernelVersion, error) {
 	}
+	patch, err := strconv.Atoi(parts[3])
+	if err != nil {
+		return nil, trace.BadParameter(
```
I would add the entire input string (that's being parsed) to all these error messages.

r0mant commented on monitoring/kernel.go:
```diff
@@ -0,0 +1,104 @@
+
+	kernelVersion, err := parseKernelVersion(release)
+	if err != nil {
+		reporter.Add(r.warningProbe(fmt.Sprintf("Failed to determine kernel version: %v.", release), err))
```
Should this be a warning? I think we should only log this as an error and not create a probe.

r0mant commented on monitoring/kernel.go:
```diff
@@ -0,0 +1,104 @@
+func (r *kernelChecker) successProbe() *pb.Probe {
+	return &pb.Probe{
+		Checker: r.Name(),
+		Detail:  "Installed linux kernel is supported.",
```
Add kernel version to the message?

r0mant commented on monitoring/kernel.go:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	if !r.isSupportedVersion(*kernelVersion) {
+		reporter.Add(r.warningProbe(fmt.Sprintf("Installed linux kernel is not supported: %v.", release), err))
```
I'd make the message less scary. `Minimum recommended kernel version is %v (%v is installed).`

r0mant commented on monitoring/kernel_test.go:
```diff
@@ -0,0 +1,68 @@
+		{
+			comment:  Commentf("Does not meet patch number requirement."),
+			expected: false,
+			version:  KernelVersion{Release: 3, Major: 10, Minor: 0, Patch: 1126},
```
Should probably be `Minor: 1` if you want to test patch component comparison.

r0mant commented on monitoring/kernel_test.go:
```diff
@@ -0,0 +1,68 @@
+			comment:  Commentf("Does not meet patch number requirement."),
+			expected: false,
+			version:  KernelVersion{Release: 3, Major: 10, Minor: 0, Patch: 1126},
+		},
```
Add a test case for 4.x.x kernel too?

bernardjkim committed:
Address PR feedback

- Fix kernel version comparison function
- String godoc
- parseKernelVersion godoc
- Additional info in failed parse errors
- Remove warning probe if parse fails
- Add kernel version to success message
- Edit warning message

bernardjkim committed:
Fix test cases

a-palchikov commented on monitoring/kernel.go:
```diff
@@ -0,0 +1,103 @@
+/*
```
If we move kernel-specific functionality into a separate file and it's still relying on Linux-specific code (eg `realKernelVersionReader`) - the source file needs to be Linux-specific and suffixed with `_linux.go` (same for test file) - otherwise we're risking build failure during publish step on darwin.

bernardjkim committed:
Edit LessThan func; Unexport kernelVersionReader

bernardjkim committed:
Rename kernel.go -> kernel_checker_linux.go

- Add default NewKernelChecker
- Update test cases

bernardjkim committed:
Move test to monitoring package

bernardjkim committed:
Unexport parseKernelVersion

bernardjkim committed:
No need to specify kernelVersionReader type
